apps/src/code-studio/components/header/EditableProjectName.jsx

```diff
@@ -64,14 +71,24 @@ class UnconnectedEditProjectName extends React.Component {
       return;
     }
 
-    dashboard.project.rename(newName, () => {
-      dashboard.header.updateTimestamp();
-      this.props.refreshProjectName();
-      this.setState({
-        savingName: false
+    dashboard.project
+      .rename(newName)
+      .then(() => {
+        dashboard.header.updateTimestamp();
+        this.props.refreshProjectName();
+        this.setState({
+          savingName: false
+        });
+        this.props.finishEdit();
+      })
+      .catch(error => {
+        if (error.responseText) {
+          const nameFailure = JSON.parse(error.responseText).nameFailure;
```

Comment:
Potential uncaught exception here from `JSON.parse` - I'm concerned that a less-expected error, something like a `503 SERVICE UNAVAILABLE` response, might leave us in a bad state.

It's potentially worth wrapping all the logic for "get the error details from this `error` object" into a helper method that provides the various try/catch and fallback behaviors we need.

Also: Should we set `savingName: false` even when save fails here?

Reply:
Alternatively, you could push the logic for converting the XHR error response to something more useful down into `project.js` - it sort of makes sense to me to hide anything ajax-specific away from this layer.